### PR TITLE
added handling of deserialization with 'kind' property and unit tests

### PR DESCRIPTION
**Pull request by ostoliarova-msft**
# Description

Made `CommunicationIdentifierConverter` convert based on `kind` property if it is set. 

[User Story 3013988](https://skype.visualstudio.com/SPOOL/_workitems/edit/3013988): [SDK][Common][AllLangs] Prefer the "kind" property when deserializing the CommunicationIdentifier - follow up

If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/implementation/converters/CommunicationIdentifierConverter.java [hunk 1]
```diff
@@ -33,6 +34,39 @@ public static CommunicationIdentifier convert(CommunicationIdentifierModel ident
...
+
+            if (kind == CommunicationIdentifierModelKind.MICROSOFT_TEAMS_USER
+                && identifier.getMicrosoftTeamsUser() != null) {
+                MicrosoftTeamsUserIdentifierModel teamsUserIdentifierModel = identifier.getMicrosoftTeamsUser();
+                Objects.requireNonNull(teamsUserIdentifierModel.getUserId());
+                Objects.requireNonNull(teamsUserIdentifierModel.getCloud());
+                Objects.requireNonNull(rawId);
+                return new MicrosoftTeamsUserIdentifier(teamsUserIdentifierModel.getUserId(),
+                    teamsUserIdentifierModel.isAnonymous())
+                    .setRawId(rawId)
```
Comment (AikoBB) [inline sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/implementation/converters/CommunicationIdentifierConverter.java:61]: in the constructor of the MicrosoftTeamsUserIdentifier the rawId is set, so I guess there is no need to reset it again

File: sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/implementation/converters/CommunicationIdentifierConverter.java [hunk 2]
```diff
@@ -33,6 +34,39 @@ public static CommunicationIdentifier convert(CommunicationIdentifierModel ident
         assertSingleType(identifier);
         String rawId = identifier.getRawId();
 
+        CommunicationIdentifierModelKind kind = identifier.getKind();
```
[the remark this answers is not in the document]
Comment (ostoliarova-msft) [reply sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/implementation/converters/CommunicationIdentifierConverter.java:37]: This file is not auto-generated. Please check comments on the top of the file, it doesn't have 'auto-generated' comment. I've discussed it with Dominik and Minnie: not all files under /implementation are auto-generated, those that don't have a respective comments, are written manually

File: sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/implementation/converters/CommunicationIdentifierConverter.java [hunk 3]
```diff
@@ -33,6 +34,39 @@ public static CommunicationIdentifier convert(CommunicationIdentifierModel ident
...
+        if (kind != null) {
+            if (kind == CommunicationIdentifierModelKind.COMMUNICATION_USER
+                && identifier.getCommunicationUser() != null) {
+                Objects.requireNonNull(identifier.getCommunicationUser().getId());
+                return new CommunicationUserIdentifier(identifier.getCommunicationUser().getId());
+            }
+
+            if (kind == CommunicationIdentifierModelKind.PHONE_NUMBER
+                && identifier.getPhoneNumber() != null) {
+                PhoneNumberIdentifierModel phoneNumberModel = identifier.getPhoneNumber();
```
[the remark this answers is not in the document]
Comment (ostoliarova-msft) [reply sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/implementation/converters/CommunicationIdentifierConverter.java:48]: Thanks! Fixed

File: sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/implementation/converters/CommunicationIdentifierConverter.java (hunk 1 above)
Comment (ostoliarova-msft) [reply sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/implementation/converters/CommunicationIdentifierConverter.java:61]: As far as I understand, only userId is set in constructor:
`    public MicrosoftTeamsUserIdentifier(String userId, boolean isAnonymous) {
        this.rawIdSet = false;
        this.cloudEnvironment = CommunicationCloudEnvironment.PUBLIC;
        if (CoreUtils.isNullOrEmpty(userId)) {
            throw new IllegalArgumentException("The initialization parameter [userId] cannot be null or empty.");
        } else {
            this.userId = userId;
            this.isAnonymous = isAnonymous;
            this.generateRawId();
        }
    }`

`setRawId` method is called from a super class and sets a raw ID
    `    public MicrosoftTeamsUserIdentifier setRawId(String rawId) {
        super.setRawId(rawId);
        this.rawIdSet = true;
        return this;
    }`

File: sdk/communication/azure-communication-callautomation/src/test/java/com/azure/communication/callautomation/CommunicationIdentifierConverterUnitTests.java [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+import com.azure.communication.callautomation.implementation.models.CommunicationIdentifierModel;
+import com.azure.communication.callautomation.implementation.models.CommunicationIdentifierModelKind;
+import com.azure.communication.callautomation.implementation.models.CommunicationUserIdentifierModel;
+import com.azure.communication.callautomation.implementation.models.MicrosoftTeamsUserIdentifierModel;
+import com.azure.communication.callautomation.implementation.models.PhoneNumberIdentifierModel;
+import com.azure.communication.common.*;
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.*;
+
+public class CommunicationIdentifierConverterUnitTests extends CallAutomationUnitTestBase {
```
[the remark this answers is not in the document]
Comment (ostoliarova-msft) [reply sdk/communication/azure-communication-callautomation/src/test/java/com/azure/communication/callautomation/CommunicationIdentifierConverterUnitTests.java:18]: Thank you! I've added more test cases.

File: sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/implementation/converters/CommunicationIdentifierConverter.java [hunk 5]
```diff
@@ -33,6 +34,39 @@ public static CommunicationIdentifier convert(CommunicationIdentifierModel ident
...
+                Objects.requireNonNull(phoneNumberModel.getValue());
+                return new PhoneNumberIdentifier(phoneNumberModel.getValue()).setRawId(rawId);
```
[the remark this answers is not in the document]
Comment (ostoliarova-msft) [reply sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/implementation/converters/CommunicationIdentifierConverter.java]: Fixed

File: sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/implementation/converters/CommunicationIdentifierConverter.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (AikoBB) [reply sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/implementation/converters/CommunicationIdentifierConverter.java:37]: Oh, I see, sorry I thought it's part of the generated code

File: sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/implementation/converters/CommunicationIdentifierConverter.java (hunk 1 above)
Comment (AikoBB) [reply sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/implementation/converters/CommunicationIdentifierConverter.java:62]: I think the value would be the same as this generating logic is approved solution and covered by tests. @petrsvihlik wdyt about setting it again?

File: sdk/communication/azure-communication-callautomation/src/test/java/com/azure/communication/callautomation/CommunicationIdentifierConverterUnitTests.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (AikoBB) [reply sdk/communication/azure-communication-callautomation/src/test/java/com/azure/communication/callautomation/CommunicationIdentifierConverterUnitTests.java:18]: the same tests scenarios for the case if the kind property is not set

File: sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/implementation/converters/CommunicationIdentifierConverter.java (hunk 1 above)
Comment (ostoliarova-msft) [reply sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/implementation/converters/CommunicationIdentifierConverter.java:62]: Currently raw ID is generated based on user ID and cloud environment only. Under some circumstances, rawId from `CommunicationIdentifierModel` could be a totally different value (like `123abc`), which could not be generated. I'm not sure what expected behaviour is but not setting a rawID explicitly here will change the previous logic.

File: sdk/communication/azure-communication-callautomation/src/test/java/com/azure/communication/callautomation/CommunicationIdentifierConverterUnitTests.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (ostoliarova-msft) [reply sdk/communication/azure-communication-callautomation/src/test/java/com/azure/communication/callautomation/CommunicationIdentifierConverterUnitTests.java:18]: Thanks, done!

File: sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/implementation/converters/CommunicationIdentifierConverter.java [hunk 6]
```diff
@@ -33,15 +34,50 @@ public static CommunicationIdentifier convert(CommunicationIdentifierModel ident
         assertSingleType(identifier);
         String rawId = identifier.getRawId();
 
+        CommunicationIdentifierModelKind kind = identifier.getKind();
+
+        if (kind != null) {
+            if (kind == CommunicationIdentifierModelKind.COMMUNICATION_USER
+                && identifier.getCommunicationUser() != null) {
+                Objects.requireNonNull(identifier.getCommunicationUser().getId());
```
[the remark this answers is not in the document]
Comment (ostoliarova-msft) [reply sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/implementation/converters/CommunicationIdentifierConverter.java]: Thanks! That's a good suggestion. Done

[the remark this answers is not in the document]
Comment (ostoliarova-msft) [reply sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/implementation/converters/CommunicationIdentifierConverter.java]: Thanks! That's a good suggestion. Done

File: sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/implementation/converters/CommunicationIdentifierConverter.java (hunk 1 above)
Comment (petrsvihlik) [reply sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/implementation/converters/CommunicationIdentifierConverter.java]: This is an exception for PhoneNumbers and MSTeamsUsers. The rawId must be set explicitly to support some legacy identifiers.

File: sdk/communication/azure-communication-callingserver/src/main/java/com/azure/communication/callingserver/implementation/converters/CommunicationIdentifierConverter.java [hunk 7]
```diff
@@ -33,30 +34,41 @@ public static CommunicationIdentifier convert(CommunicationIdentifierModel ident
...
-                .setRawId(rawId)
-                .setCloudEnvironment(CommunicationCloudEnvironment
-                    .fromString(teamsUserIdentifierModel.getCloud().toString()));
+            return getMicrosoftTeamsUserIdentifier(identifier, rawId);
         }
 
-        Objects.requireNonNull(rawId);
+        Objects.requireNonNull(rawId, "'RawID' of the CommunicationUserIdentifierModel cannot be null.");
         return new UnknownIdentifier(rawId);
     }
```
Comment (petrsvihlik) [inline sdk/communication/azure-communication-callingserver/src/main/java/com/azure/communication/callingserver/implementation/converters/CommunicationIdentifierConverter.java:73]: pls refactor according to https://github.com/Azure/azure-sdk-for-net/pull/31627/files#r1000528967

**Review by petrsvihlik — CHANGES_REQUESTED** [review]
and pls bump the versions of touched pkgs

File: sdk/communication/azure-communication-callingserver/src/main/java/com/azure/communication/callingserver/implementation/converters/CommunicationIdentifierConverter.java (hunk 7 above)
Comment (ostoliarova-msft) [reply sdk/communication/azure-communication-callingserver/src/main/java/com/azure/communication/callingserver/implementation/converters/CommunicationIdentifierConverter.java]: Done

File: sdk/communication/azure-communication-callingserver/CHANGELOG.md [hunk 8]
```diff
@@ -16,6 +16,7 @@ This is the first version of the restart of Azure Communication Service Calling
 - Call recording.
 - Get, add and remove participants from the call.
 - Recording download apis.
+- Added support for an optional `kind` property for a converter of `CommunicationIdentifierModel`.
```
Comment (petrsvihlik) [inline sdk/communication/azure-communication-callingserver/CHANGELOG.md:19]: `CommunicationIdentifierModel` is not extracted as a part of the public API, we shouldn't mention it here. 
```suggestion
- Optimized the (de)serialization of the types derived from the `CommunicationIdentifier`.
```

**Review by petrsvihlik — CHANGES_REQUESTED** [review]

File: sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/implementation/converters/CommunicationIdentifierConverter.java [hunk 9]
```diff
@@ -32,31 +33,37 @@ public static CommunicationIdentifier convert(CommunicationIdentifierModel ident
...
         }
 
-        if (identifier.getPhoneNumber() != null) {
-            PhoneNumberIdentifierModel phoneNumberModel = identifier.getPhoneNumber();
-            Objects.requireNonNull(phoneNumberModel.getValue());
-            return new PhoneNumberIdentifier(phoneNumberModel.getValue()).setRawId(rawId);
+        if (kind == CommunicationIdentifierModelKind.PHONE_NUMBER
+            && identifier.getPhoneNumber() != null) {
+            String phoneNumber = identifier.getPhoneNumber().getValue();
+            Objects.requireNonNull(phoneNumber, "'PhoneNumber' of the CommunicationUserIdentifierModel cannot be null.");
```
Comment (petrsvihlik) [inline sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/implementation/converters/CommunicationIdentifierConverter.java:48]: `CommunicationUserIdentifierModel` doesn't contain a `PhoneNumber`. It should either be `CommunicationIdentifierModel`, or its submodel `PhoneNumberIdentifierModel`.

File: sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/implementation/converters/CommunicationIdentifierConverter.java [hunk 10]
```diff
@@ -32,31 +33,37 @@ public static CommunicationIdentifier convert(CommunicationIdentifierModel ident
...
 
-        if (identifier.getMicrosoftTeamsUser() != null) {
+        if (kind == CommunicationIdentifierModelKind.MICROSOFT_TEAMS_USER
+            && identifier.getMicrosoftTeamsUser() != null) {
             MicrosoftTeamsUserIdentifierModel teamsUserIdentifierModel = identifier.getMicrosoftTeamsUser();
-            Objects.requireNonNull(teamsUserIdentifierModel.getUserId());
-            Objects.requireNonNull(teamsUserIdentifierModel.getCloud());
-            Objects.requireNonNull(rawId);
+            Objects.requireNonNull(teamsUserIdentifierModel.getUserId(), "'UserID' of the CommunicationUserIdentifierModel cannot be null.");
+            Objects.requireNonNull(teamsUserIdentifierModel.getCloud(), "'Cloud' of the CommunicationUserIdentifierModel cannot be null.");
```
Comment (petrsvihlik) [inline sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/implementation/converters/CommunicationIdentifierConverter.java:57]: same here

File: sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/implementation/converters/CommunicationIdentifierConverter.java [hunk 11]
```diff
@@ -32,31 +33,37 @@ public static CommunicationIdentifier convert(CommunicationIdentifierModel ident
...
+            Objects.requireNonNull(rawId, "'RawID' of the CommunicationUserIdentifierModel cannot be null.");
             return new MicrosoftTeamsUserIdentifier(teamsUserIdentifierModel.getUserId(),
                 teamsUserIdentifierModel.isAnonymous())
                 .setRawId(rawId)
                 .setCloudEnvironment(CommunicationCloudEnvironment
                     .fromString(teamsUserIdentifierModel.getCloud().toString()));
         }
 
-        Objects.requireNonNull(rawId);
+        Objects.requireNonNull(rawId, "'RawID' of the CommunicationUserIdentifierModel cannot be null.");
```
Comment (petrsvihlik) [inline sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/implementation/converters/CommunicationIdentifierConverter.java:66]:
```suggestion
        Objects.requireNonNull(rawId, "'RawID' of the CommunicationIdentifierModel cannot be null.");
```

**Review by petrsvihlik — CHANGES_REQUESTED** [review]

**Review by petrsvihlik — APPROVED** [review]

**Review by AikoBB — APPROVED** [review]

**Pull request merged**